### PR TITLE
[Flight] Prefix Replayed Console Logs with a Badge

### PR DESCRIPTION
**Pull request by sebmarkbage**
Builds on top of #28384.

This prefixes each log with a badge similar to how we badge built-ins like "ForwardRef" and "Memo" in the React DevTools. The idea is that we can add such badges in DevTools for Server Components too to carry on the consistency.

This puts the "environment" name in the badge which defaults to "Server". So you know which source it is coming from.

We try to use the same styling as the React DevTools. We use light-dark mode where available to support the two different color styles, but if it's not available I use a fixed background so that it's always readable even in dark mode.

In Terminals, instead of hard coding colors that might not look good with some themes, I use the ANSI color code to flip background/foreground colors in that case.

In earlier commits I had it on the end of the line similar to the DevTools badges but for multiline I found it better to prefix it. We could try various options tough.

In most cases we can use both ANSI and the `%c` CSS color specifier, because node will only use ANSI and hide the other. Chrome supports both but the color overrides ANSI if it comes later (and Chrome doesn't support color inverting anyway). Safari/Firefox prints the ANSI, so it can only use CSS colors.

Therefore in browser builds I exclude ANSI.

On the server I support both so if you use Chrome inspector on the server, you get nice colors on both terminal and in the inspector.

Since Bun uses WebKit inspector and it prints the ANSI we can't safely emit both there. However, we also can't emit just the color specifier because then it prints in the terminal. https://github.com/oven-sh/bun/issues/9021 So we just use a plain string prefix for now with a bracket until that's fixed.

Screen shots:

Screenshot 2024-02-21 at 12 56 02 AM
Screenshot 2024-02-21 at 12 56 24 AM
Screenshot 2024-02-21 at 12 57 10 AM
Screenshot 2024-02-21 at 12 57 34 AM
Screenshot 2024-02-21 at 12 58 23 AM
Screenshot 2024-02-21 at 12 58 56 AM

### CODE REVIEW

**Comment by gaearon** [comment]
The badges are sick. 🔥

File: packages/react-client/src/ReactFlightClientConsoleConfigBrowser.js [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    case 'table': {
+      // These methods cannot be colorized because they don't take a formatting string.
+      // eslint-disable-next-line react-internal/no-production-logging
+      console[methodName].apply(console, args);
+      return;
+    }
+    case 'assert': {
+      // assert takes formatting options as the second argument.
+      offset = 1;
+    }
```
Comment (hoxyq) [inline packages/react-client/src/ReactFlightClientConsoleConfigBrowser.js:42]: Do we allowlist these methods somewhere? I wonder what the behaviour will be if some non-standard console methods are called, like `console.timeStamp` or `console.createTask`?

Comment (hoxyq) [reply packages/react-client/src/ReactFlightClientConsoleConfigBrowser.js:42]: I think I found the answer - https://github.com/facebook/react/pull/28384/files#r1495197550

**Review by hoxyq — APPROVED** [review]
Reviewed 6 commits at the top, basically what is stacked. Cool stuff.

File: packages/react-client/src/ReactFlightClientConsoleConfigBrowser.js (hunk 1 above)
Comment (sebmarkbage) [reply packages/react-client/src/ReactFlightClientConsoleConfigBrowser.js:42]: Yes, we only patch a subset. The list and what was excluded is commented on here:

https://github.com/facebook/react/pull/28384#discussion_r1495197550

**Comment by rickhanlonii** [comment]
This is sick

**Pull request merged**